python_benchmarks/test_rope.py

```diff
@@ -0,0 +1,184 @@
+import pytest
+from nvfuser import FusionDefinition, DataType
+from .core import run_benchmark, clear_cuda_cache
+import torch
+
+
+# Mimic the Hugging Face implementation:
+# https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L216
+def rope_with_cat_fusion(
+    fd: FusionDefinition,
+    batch_size: int,
+    seq_len: int,
+    num_heads: int,
+    features_per_head: int,
+) -> None:
+    q = fd.define_tensor(
+        shape=[batch_size, seq_len, num_heads, features_per_head],
+        dtype=DataType.BFloat16,
+    )
+    cos = fd.define_tensor(
+        shape=[seq_len, features_per_head],
+        dtype=DataType.BFloat16,
+    )
+    sin = fd.define_tensor(
+        shape=[seq_len, features_per_head],
+        dtype=DataType.BFloat16,
+    )
+
+    q = fd.ops.permute(q, dims=[0, 2, 1, 3])
+    q_real = fd.ops.slice(
+        q,
+        start_indices=[0, 0, 0, 0],
+        end_indices=[batch_size, num_heads, seq_len, features_per_head // 2],
+        strides=[1, 1, 1, 1],
+    )
+    q_image = fd.ops.slice(
+        q,
+        start_indices=[0, 0, 0, features_per_head // 2],
+        end_indices=[batch_size, num_heads, seq_len, features_per_head],
+        strides=[1, 1, 1, 1],
+    )
+
+    # nvFuser has problems generating negation for bfloat.
+    q_image = fd.ops.cast(q_image, dtype=DataType.Float)
+    q_image = -q_image
+    q_image = fd.ops.cast(q_image, dtype=DataType.BFloat16)
+
+    q_rotated = fd.ops.cat([q_image, q_real], dim=-1)
+
+    cos = fd.ops.broadcast_in_dim(
+        cos, shape=[1, 1, seq_len, features_per_head], broadcast_dims=[2, 3]
+    )
+    sin = fd.ops.broadcast_in_dim(
+        sin, shape=[1, 1, seq_len, features_per_head], broadcast_dims=[2, 3]
+    )
+
+    out = q * cos + q_rotated * sin
+    out = fd.ops.cast(out, DataType.BFloat16)
```

Comment:
Is this needed? All the inputs are `BFloat16`.

Reply:
we don't do fp16 math. I imagine without the explicit cast, we'll be generating fp32 output and that's gonna be a surprise.

Reply:
Oh right. OOC, is there a difference performance-wise between explicitly casting all the inputs to float vs letting nvFuser implicitly cast inputs? (Context: we explicitly cast inputs to float in other benchmark fusions)

Reply:
I have no opinions between the two at this moment. In nvFuser, implicit casts are done where they are needed (e.g. for mul but not for reshape), so using implicit cast can be more performant in some situations. On the other hand, I doubt everyone on the team knows where implicit casts are inserted, so using explicit casts probably makes code easier to read.